### PR TITLE
[Refactor] Particle variables and functions handling

### LINKED ISSUE #637: Refactor particle to handle different types
## Summary

Refactor `Particle` data to have the generic functions with the ability to fetch and modify generic scalar and vector properties, map, and interpolate these properties to nodes. Add a separate implementation functions to operate on these properties of the particle. These functions could be grouped under the namespace of solid particle functions, fluid-particle function, and mixed particle functions. The particle agnostic design of functions would allow for mix-and-matching different functions in the Solver class, without having to deal with multiple-inheritance or repetition.

## Motivation

Refactoring and separating the Particle data from the implementation of algorithm enables handling different particle data types, without having to have a lot of virtual functions in the ParticleBase and other subsequent derived particles. Addresses the comments raised in RFCs #633 and #634. For example, deriving `FluidParticle` from `Particle` would result in extra data needed for `FluidParticle`, which is not needed in the case of `Particle` conversely, `FluidParticle` won't need certain `Particle` data. However, the important consideration is that the functions of a `FluidParticle`s have to be virtualized in the abstract `ParticleBase` and every time a new class is derived, more abstract functions have to be added to the base and other existing derived classes. These functions, which are not useful, must, therefore, throw an error. The Abstract Factory design with inconsistent derived functions is inherently prone to errors when a developer fails to add functions to all the derived classes. This RFC proposes an alternate way of deriving particle data and keeping the implementation separate and reusable. Separation of implementation allows for the composition of different functions in the Solver class.

## Design Detail
Particle class has a generic ordered map of scalar and vector properties. An open addressing scheme will be used to store the map properties, that the data stored in here will remain continuous in memory. These properties can be accessed using `scalar_property` and `vector_property`. `Particle` types will still control the initialization of these different properties, i.e., derived classes will be used to create new types of Particles, but all these classes will have a consistent functional interface. This avoids the problem of too many virtual functions. The `ParticleBase` will be altered to have the following. A consistent interface allows transferring particles across MPI ranks.

```c++
// Scalar properties
tsl::ordered_map<std::string, double> scalar_properties_;
tsl::ordered_map<std::string, Eigen::Matrix<double, Tdim, 1>> vector_properties_;
```
Scalar and vector properties can be set and returned using generic functions. 
```c++
void assign_property(const std::string& property, double value);
double property(const std::string& property) const;
```

Nodes will also have a similar property pool structure, which will allow us to write generic map and interpolation functions. 

```c++
//! Map scalar properties to nodes
template <unsigned Tdim>
void mpm::Particle<Tdim>::map_scalar_properties_to_nodes(const std::string& property) noexcept {
  auto value = scalar_properties_.at("property");
  // Check if scalar property is not max
  assert(value != std::numeric_limits<double>::max());

  // Map mass and momentum to nodal property taking into account the material id
  for (unsigned i = 0; i < nodes_.size(); ++i) {
    value = value * shapefn_[i];
    nodes_[i]->update_property(true, property, value);
  }
}
```

This property pool structure provides a generic interface to assign and get properties. We can now define a collection of functions under a namespace, such as `particle::fluid` to define functions that operate on fluid properties. The initialization of the property pool for the Particle will be done by the derived `Particle` class. Each particle type implementation will have access to the property pool data. 

```c++
  // Assign projection parameter
  template <unsigned Tdim>
  void mpm::particle::fluid::projection_parameter(std::shared_ptr<mpm::ParticleBase<Tdim>> particle, double value) {
    particle->assign_scalar_property("projection", value);
  } 
```

In the custom solver for the fluid particle, we'll have:

```c++
// Iterate over all the particles and assign a projection parameter
double value = 2.4;
mesh->iterate_over_particles(
       [&value](std::shared_ptr<mpm::ParticleBase<Dim>> ptr) {
          return mpm::particle::fluid::projection_parameter<Dim>(ptr, value);
 });
```

To operate on a specific set of particle types, we could use `iterate_over_particles_with_predicate` functionality that can filter particles with a particular type.

With this approach, a solver class can build on different types of Particle data and functions pooled from multiple namespaces.

The proposal is to start moving new `Particle` types to use this design, and the existing solid particle will be refactored in phases.

## Drawbacks

The implementation operating on the derived data types is separated from the actual data, which is good in some ways, but also it might be seen as breaking encapsulation. Considering the fact that if something owns a `Particle` type, then it should be able to modify the data in the `Particle`. This separation of data and implementation may give some developers a false feeling that their implementation could work on all `Particle` types, however, as soon as the code is run or compiled with debug mode, it will be known that the data type is unsupported and cannot be used. The phase refactoring of Solid Particles will break a lot of existing code, however, it will minimize the number of `update_nodes` and `interpolate_from_nodes` design.

## Rationale and Alternatives

> Advantages

* Allows for future composition in the solver class with multiple Particle types, multiple `particle::fluid` functions can be defined and used in different solver classes.

* Avoids complex inheritance schemes, for example, `TwoPhaseParticle` doesn't have to redefine functions in `fluid` or `solid` when derived from only one type.

> Other designs considered

* FactoryDesign (Abstract Class): This results in too many virtual functions, an addition of a new function in one of the derived classes affects all existing `Particle` classes.

* BridgeDesign: Requires a class for the implementation that takes the `ParticleData`, same as Abstract class, results in virtual functions pushed to implementation.

* std::variant: Good alternative to removing abstract base classes. However, may not be a good alternative, as we still need to separate the functions from the data. Also, different template instantiation is not possible.

* Super functions: Use more generic super functions which are public member functions and specialized functions for each derived type are private member functions. This design avoids non-pure virtual functions, however, limits the interface functions so it is not flexible and writing test functions are difficult.

> What is the impact of not doing this?

* Consistent interface cannot be achieved on different particles and will have to refactor existing classes every time a new function is added. 

* Consistent interface helps with MPI data transfer.

## Prior Art

Most large codes avoid a lot of derived classes and Abstract class design with inconsistent interfaces across derived classes.

> Dealii uses an extreme case of all properties being double and provide particle handle. [Dealii Particle](https://github.com/dealii/dealii/blob/master/include/deal.II/particles/particle.h#L85)

> Uintah also uses a generic datawarehouse to store properties.

> [No Paradigm programming](https://www.youtube.com/watch?v=t_0JhHVh_Kg)

> [Free your functions](https://www.youtube.com/watch?v=WLDT1lDOsb4)

## Unresolved questions

> What parts of the design do you expect to resolve through the RFC process before this gets merged?

New Particle types will follow this new design, while the existing solid particle has to be slowly refactored.

> What related issues do you consider out of scope for this RFC that could be addressed in the future independently of the solution that comes out of this RFC?

The refactoring of solver class to make it more composable will be discussed in a separate RFC.

> Unresolved questions

How to handle MPI data transfer with an arbitrary number of particle data types? Since the `ParticleBase` will be derived for each type of Particle, we will have enough information to transfer particles. Suggest creating serialization and deserialization schemes. Additionally, since the number of active `Particle` types are going to be a handful at most, we can iterate through different types when doing MPI halo exchange.

Still undecided if we need to derive a new `Particle` for each type, so the advantage of this method is that the derived `Particle` type knows its variables and is easy for instantiation. However, it creates additional vtables. On the other hand, if we move the variable instantiation outside of the `Particle`, we don't have to derive new types, but initialization and MPI transfer may become harder?

## Changelog
[Edit #1]: Update `iterate_over_mesh` function and formatting.
[Edit #2]: Add std::variant as alternatives and prior art
[Edit #3]: Unresolved question on deriving new Particle types
[Edit #4]: Discussion on superfunctions as an alternative

### PR DESCRIPTION
**Pull request by kks32**
**Describe the PR**
This is an alternative implementation of particle refactor. 

After some more consideration, I am wondering about the possibility to derive particle types, but only keep the common particle functions within the class, and merge all independent/specialized functions to be outside the class. This will help to initialize the classes with the right types, and also in MPI transfer, as the type is known. We derived data of individual Particle types but keep specialized particle functions separately. Maybe the particle functions may have to stay in the same namespace.

**Related Issues/PRs**
#637 and #650 

**Additional context**
Specialized functions will be still separate.

Following are the run-times for 3D hydrostatic column 10k steps on Stampede2

| Branch                   | TACC runtime (s) |
|--------------------------|-------------------|
| develop                  | 742.6            |
| prop map + free fn | 1280             |
| vector + free fn map       | 865             |
| flatmap + free fn       | 972            |

### CODE REVIEW

File: include/node.tcc
Comment (kks32) [inline include/node.tcc]:
```suggestion
  // Update/assign value
```

File: include/node.tcc
Comment (kks32) [inline include/node.tcc:114]: Should we move this check to `update_vector_property`

File: include/particles/particle.h
Comment (kks32) [inline include/particles/particle.h:337]: I thought we were going to keep volume inside particle, no?

Comment (bodhinandach) [reply include/particles/particle.h:337]: I thought there might be a chance to have `liquid_volume` too in the future, so thinking to have them inside `properties`.

File: tests/particle_test.cc
Comment (ezrayst) [inline tests/particle_test.cc]: Any thoughts why change the value of the test?

Comment (bodhinandach) [reply tests/particle_test.cc]: That's just to check the return and assign values to mass, no different. I will revert to minimize the change. Thanks for pointing that out.

File: tests/particle_test.cc
Comment (ezrayst) [inline tests/particle_test.cc:682]: Why is this deleted?

File: tests/particle_test.cc
Comment (ezrayst) [inline tests/particle_test.cc:775]: There is no `try` in `compute_mass` actually. Should we still require this?

File: tests/particle_test.cc
Comment (ezrayst) [inline tests/particle_test.cc:772]: Same as above

File: tests/particle_test.cc
Comment (ezrayst) [reply tests/particle_test.cc:775]: Sorry, there are `try` within the functions in this function. If any would `throw`, it would be caught here right?

File: tests/particle_test.cc
Comment (ezrayst) [inline tests/particle_test.cc]: Same as the other one, but it's really up to you, I thought there are more fundamental reasons to change

File: tests/particle_test.cc
Comment (bodhinandach) [reply tests/particle_test.cc:682]: That's commented test. Tbh, there are many of them found all related to assert. I would prefer to clean them up. What do you think @kks32?

Comment (kks32) [reply tests/particle_test.cc:682]: I think catch has a `REQUIRE_THROW` with a specific exception, we could test against those

**Comment by bodhinandach** [comment]
Dear all, I think this refactoring branch is ready for review. Could you all please some time to take a look at it? We can discuss it for sure in the following developer meeting. Here are some notes to ease the checking process:
- The main changes are located mainly in `particle` and `node` class with the addition of `mpm_properties` and `particle_functions`. 
- Some functions in `particle` are moved out to `particle_functions`.
- Assign (or update) and return variables in nodes and particles are kept as before.
- In particle, tensorial properties are still kept as before, we will refactor these in another PR, if necessary.
- After a discussion with @thiagordonho, the contact functions in particle class are still kept as it is in this PR. We need to refactor how nodal_properties are stored and hopefully coming with similar approach to this. This will also be in a separate PR.
- Please refer to issue #637 for further details. 

As this change is breaking for all the development, could you please test it to run some of your simulations and let us know if you encounter any issue. Many many thanks in advance for the help. Any comments and questions are appreciated.

File: tests/particle_test.cc
Comment (bodhinandach) [reply tests/particle_test.cc:775]: yes

File: tests/particle_test.cc
Comment (bodhinandach) [reply tests/particle_test.cc:682]: @kks32 Should I revert the deleted comment and change to `REQUIRE_THROW`? Even though it is not tested here, I think it has been called in some other part of the test with `REQUIRE_NOTHROW`.

File: .circleci/config.yml
Comment (bodhinandach) [inline .circleci/config.yml]: A small reminder to revert this back before merging.

**Comment by tianchiTJ** [comment]
Why `initialise_particle(const HDF5Particle& particle, const std::shared_ptr<Material<Tdim>>& material)` is defined as virtual function in `particle.h`?

**Comment by bodhinandach** [comment]
> Why `initialise_particle(const HDF5Particle& particle, const std::shared_ptr<Material<Tdim>>& material)` is defined as virtual function in `particle.h`?

Good point. I will remove it, we should either use `virtual` or `override`, but not both in one go.

**Comment by tianchiTJ** [comment]
> > Why `initialise_particle(const HDF5Particle& particle, const std::shared_ptr<Material<Tdim>>& material)` is defined as virtual function in `particle.h`?
> 
> Good point. I will remove it, we should either use `virtual` or `override`, but not both in one go.

This one also exists in the `develop` branch

**Comment by bodhinandach** [comment]
@tianchiTJ the most recent commit should fix, thanks

File: include/particles/particle.h
Comment (tianchiTJ) [inline include/particles/particle.h:127]: The current version define all simple return value functions in the derive class. How can I define such function for two phase? Such as `liquid_mass()` `porosity()` `liquid_material()`

Comment (bodhinandach) [reply include/particles/particle.h:127]: If it is inside your `twophase_particle`, you can create a private function to ease the calling process, for instance:
```
private:
double pore_pressure() const {
    return this->scalar_property(mpm::properties::Scalar::PorePressure);
}
```
However, to call it from the `twophase_particle_functions`, testings, or outside files, you can use:
```
particle->scalar_property(mpm::properties::Scalar::PorePressure);
```

**Comment by jgiven100** [comment]
I'm getting matching outputs running a PEA simulation on `develop` and `refactor/particles` 👍

File: include/particles/particle.h
Comment (tianchiTJ) [reply include/particles/particle.h:127]: Yes, I need it in the `twophase_particle_functions`. I can do this way, but it will be not identical to the single phase format.

Comment (bodhinandach) [reply include/particles/particle.h:127]: I understand. That should be ok.

Comment (bodhinandach) [reply include/particles/particle.h:127]: If you use the variable multiple times, maybe you want to define it in the beginning of the function, such that:
`const auto pore_pressure = particle->scalar_property(mpm::properties::Scalar::PorePressure);`
and you can use it multiple times.

Comment (tianchiTJ) [reply include/particles/particle.h:127]: This also means a particle pointer should be set as an argument of the functions.

Comment (bodhinandach) [reply include/particles/particle.h:127]: Yes, you are right, the same as other functions in `particle_functions`.

File: include/particles/particle.h
Comment (tianchiTJ) [inline include/particles/particle.h:364]: Why not put the `traction_` into the `vector_properties_`?

File: include/particles/particle_functions.tcc
Comment (tianchiTJ) [inline include/particles/particle_functions.tcc:16]: Could we move this assign density function into the `assign_material`? Because for the twophase, there is no pointer for the `liquid_material` in the `twophase_particle_functions.tcc`

File: include/particles/particle_functions.tcc
Comment (tianchiTJ) [inline include/particles/particle_functions.tcc]: I think the `map_scalar_property_nodes` has a problem to be used for two phase. The liquid mass is called `LiquidMass` in the particle properties, but `Mass` at liquid_phase in the node properties. That's conflict.

Comment (bodhinandach) [reply include/particles/particle_functions.tcc]: Use the overload function with input value as your `LiquidMass`
https://github.com/cb-geo/mpm/blob/164491fd7f5c7ef09536405aeb2ae0d5fd8947ea/include/particles/particle_base.tcc#L50-L53

File: include/node.h
Comment (thiagordonho) [inline include/node.h:76]: `//! \param[in] property Name of the property to return`

File: include/node_base.h
Comment (thiagordonho) [inline include/node_base.h:111]:
```suggestion
  //! Return property at a given node for a given phase
  //! \param[in] property Name of the property to return
```

File: include/node_base.h
Comment (thiagordonho) [inline include/node_base.h:95]:
```suggestion
  //! Return property at a given node for a given phase
  //! \param[in] property Name of the property to return
```

File: include/node.h
Comment (thiagordonho) [inline include/node.h:104]:
```suggestion
  //! Return property at a given node for a given phase
  //! \param[in] property Name of the property to return
```

File: include/node.h
Comment (thiagordonho) [inline include/node.h:88]:
```suggestion
  //! Return property at a given node for a given phase
  //! \param[in] property Name of the property to return
```

File: include/node.tcc
Comment (kks32) [inline include/node.tcc:268]: Don't we have pressure as a state variable?

Comment (bodhinandach) [reply include/node.tcc:268]: Yes, that's on `particle`, but on `node` we store it as `mpm::properties`.

**Review by thiagordonho — DISMISSED** [review]

File: include/particles/particle_functions.tcc
Comment (bodhinandach) [reply include/particles/particle_functions.tcc:16]: Thanks, I will address it in a future PR.

File: include/particles/particle.h
Comment (bodhinandach) [reply include/particles/particle.h:364]: As discussed, BC will also be in a future PR.

**Review by tianchiTJ — COMMENTED** [review]
How could I call the two phase variable, such as `liquid_material` in the `twophase_particle_functions.tcc`, which is neither a scalar nor a vector?

**Review by tianchiTJ — DISMISSED** [review]
Problem with `liquid_material` will be addressed in a follow up PR @bodhinandach

**Review by bodhinandach — DISMISSED** [review]

**Comment by ezrayst** [comment]
Anyways, how is the performance of using ordered map on different scalar and vector properties? Any comparison done yet?

**Comment by kks32** [comment]
The 3D benchmarks of hydrostatic column shows this branch is about 1.5 - 2x slower. 

develop: 615931 ms
refactor-particles: 980017 ms

TACC develop: 742655 ms
TACC refactor-particles: 1280757 ms

**Comment by kks32** [comment]
> Anyways, how is the performance of using ordered map on different scalar and vector properties? Any comparison done yet?

We have to look at that now separately to see if that gives us any benefit.

**Comment by bodhinandach** [comment]
@kks32 I created a partially revert branch `refactor/particles_2` to move back the functions in `particle_functions.tcc` back to `particles`. Just want to see whether the slowdown is caused by restructuring the properties or moving the function out to a separate file with passing pointer argument. I am re-running the same 3D hydrostatic benchmark locally.

**Comment by kks32** [comment]
Looks like the map function is the reason

| Branch                   | Local runtime (s) | TACC runtime (s) |
|--------------------------|-------------------|------------------|
| develop                  | 615.9             | 742.6            |
| particle (prop map + free fn) | 980               | 1280             |
| particle2 (only prop map)     | 872.9             | 1250             |

**Comment by kks32** [comment]
image
Definitely the map lookup is the reason for the slowness

**Comment by bodhinandach** [comment]
@kks32 Thanks for checking! I was worried about this before. Any thoughts on changing the data structure?

**Comment by kks32** [comment]
I'm trying out a few different data structures, I'll let you know once I have the results.

**Comment by kks32** [comment]
I tried a few different structures:


| Branch                   | TACC runtime (s) |
|--------------------------|-------------------|
| develop                  | 742.6            |
| prop map + free fn | 1280             |
| only prop map       | 1250             |
| only robin map       | 1012            |
| only vector       | 850             |

The vector implementation seems close to develop. Let me try vector implementation with free functions

**Comment by kks32** [comment]
> I tried a few different structures:
> Branch 	TACC runtime (s)
> develop 	742.6
> prop map + free fn 	1280
> only prop map 	1250
> only robin map 	1012
> only vector 	850
> 
> The vector implementation seems close to develop. Let me try vector implementation with free functions

Vector with free functions is 865s.

**Comment by bodhinandach** [comment]
Thanks @kks32, that makes it 1.16 times slower than the original. Are you checking the OpenMP/MPI performance as well? 
As the current refactor is created purposely for fluid and two-phase implementation, do you want me to check the vector one with the navier-stokes branch too?

**Comment by kks32** [comment]
Vector implementation profiling

profile

File: include/particles/particle.tcc
Comment (bodhinandach) [inline include/particles/particle.tcc]: @kks32 I think this could be a problem. Currently, in `mpm_properties`, `Mass`, `Volume`, and `MassDensity` are defined with index 0, 1, and 2, which is safe for this case. But this will be an error if we want to add another variable which enum associated index is larger than the `size` that you specified in `scalar_properties_.resize(size);`. It will be a memory access error.

File: include/particles/particle.tcc
Comment (bodhinandach) [inline include/particles/particle.tcc]: The same here with `Displacement` and `Velocity`...

**Review by bodhinandach — CHANGES_REQUESTED** [review]

File: include/node.tcc
Comment (bodhinandach) [inline include/node.tcc]: `Displacement` is not needed in node, but I know why we need to append it here to avoid access error. Does that means for every variable added in `mpm_properties`, we need to arrange and allocate all other particles? I don't think that's efficient storage-wise.

**Comment by kks32** [comment]
Profiling of develop

image

**Comment by kks32** [comment]
The problem with vector is like you said, we may have to fill everything otherwise we have an access issue.

**Comment by kks32** [comment]
Profiling for robin hasmap

image

**Comment by kks32** [comment]
| Branch                   | TACC runtime (s) |
|--------------------------|-------------------|
| develop                  | 742.6            |
| prop map + free fn | 1280             |
| vector + free fn map       | 865             |
| AssocVec + free fn       | 972            |

Profiling of develop
image

Profiling of robinhashmap
image

Profiling of vector
image

Profiling of AssocVector
image

**Comment by kks32** [comment]
Profiling of AssocVector

image

**Comment by kks32** [comment]
After considering the speed-ups and different approaches, the use of scalar and vector properties still is 30% slower than using private variables. The advantage of scalar/vector properties is we'll have fewer functions and less repetition of code. On the other hand, defining private variables and deriving particle classes has more function definitions and repeating code, but is faster. I'd like to propose a vote, use emojis to vote between different options, please use comments below if you like to elaborate on your thoughts. @bodhinandach @tianchiTJ @thiagordonho @cw646 @ezrayst @jgiven100 @WeijianLIANG

1. Use derived particle types with private variables (:rocket:)
2. Solid particles will remain as is, while we add scalar and vector properties to derived particle types (:smile:) PR #673 
3. Cleaner but slower design with free functions and scalar/vector properties for all particle types (:heart:) PR #654
4. Go back to the drawing board (:eyes:)

**Comment by WeijianLiang** [comment]
Thanks Krishna, I would prefer the use of scalar and vector properties which has a more clear structure.

**Comment by kks32** [comment]
@WeijianLIANG is that option 2 or 3? Could you please vote with a reaction emoji :smile:  for option 2 and :heart:  for option 3. Thanks!

**Comment by bodhinandach** [comment]
@kks32 I am voting for 3 in the long term as we found other methods to improve the speed, but as for now, I think option 2 is the best.

**Comment by yliang-sn** [comment]
> After considering the speed-ups and different approaches, the use of scalar and vector properties still is 30% slower than using private variables. The advantage of scalar/vector properties is we'll have fewer functions and less repetition of code. On the other hand, defining private variables and deriving particle classes has more function definitions and repeating code, but is faster. I'd like to propose a vote, use emojis to vote between different options, please use comments below if you like to elaborate on your thoughts. @bodhinandach @tianchiTJ @thiagordonho @cw646 @ezrayst @jgiven100 @WeijianLIANG
> 
> 1. Use derived particle types with private variables (🚀)
> 2. Solid particles will remain as is, while we add scalar and vector properties to derived particle types (😄)
> 3. Cleaner but slower design with free functions and scalar/vector properties for all particle types (❤️)
> 4. Go back to the drawing board (👀)

For 2, are we just dealing with the scalar and vector properties to the derived particle types? If we have other types of quantities, do we still use the private variables?

**Comment by kks32** [comment]
@yliang-sn everything will be held within scalar/vector/tensor/boolean

**Comment by yliang-sn** [comment]
> @yliang-sn everything will be held within scalar/vector/tensor/boolean

Can I vote, too? If the private variables have the best efficiency, I prefer to 1. The structure is also clear enough to be read and extended. However, 2 is also fine for me.

**Comment by kks32** [comment]
@yliang-sn yes, I couldn't find your handle, please vote.

**Comment by kks32** [comment]
@thiagordonho @tianchiTJ @cw646 @WeijianLIANG could you please use emoji reactions to vote on this comment please: https://github.com/cb-geo/mpm/pull/654#issuecomment-662429917

**Comment by ezrayst** [comment]
Both @kks32 and @bodhinandach have been working hard on this issue so appreciate your work. I think it has been about 2 months now. 

However, I feel that we need a unanimous decision on this refactor, or at least a 2/3 decision. The reason being is I don't want us to be in this position within 6 months where we are thinking to refactor again. I think the timeline on this issue has shown that this is a lot of work and everyone's progress is halted. *Personally, if we are not even sure on (2) or (3), or anyone still has reservation on them, wouldn't it better to keep it as is (1)?* Yes, `particle_base` will be messy because of the many functions from `two_phase` particles, but at least we get the speedup. Yes we might need to refactor in the future, but isn't that the assumption that we will refactor again? Lastly, if you are sure for (2), might as well go for (3) so we don't come back to it again later. So my vote will be (3) now!

**Comment by WeijianLiang** [comment]
@kks32  Thanks, I also vote for (2)

**Comment by bodhinandach** [comment]
After considering @ezrayst comment, I changed my vote to option (3). I know that the code will be slower, but the code is structurally way better and tidier. Also, if currently by moving to (3) causing us to be 30% slower than develop, I think we should rethink any other way to improve the speed. For instance, PR #677 has already improved the speed by 4% (thanks to @kks32). I think we can regain the current performance in short time by improving efficiency in other parts.

**Comment by yliang-sn** [comment]
After talking with Ezra and Nanda, I think i will change to vote for 3 if we can improve the efficiency in other aspects.

**Comment by kks32** [comment]
Other containers like Abseil, folly, vector of vectors, static_vector, and small_vector implementations all gave slower or almost identical performance as the flat_map. Due to the 30% reduction in performance we are closing this issue. https://trello.com/c/fGF3MrKg/110-plan-for-next-week-notes

**Pull request closed without merging**
